// html-validate/build/rules/attribute-boolean-style.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const rule_1 = require("../rule");
const defaults = {
    style: 'omit',
};
class AttributeBooleanStyle extends rule_1.Rule {
    constructor(options) {
        super(Object.assign({}, defaults, options));
        this.hasInvalidStyle = parseStyle(this.options.style);
    }
    setup() {
        this.on('dom:ready', (event) => {
            const doc = event.document;
            doc.visitDepthFirst((node) => {
                const meta = node.meta;
                if (!meta || !meta.attributes)
                    return;
                for (const [key, attr] of Object.entries(node.attr)) {
                    if (!this.isBoolean(key, meta.attributes))
                        continue;
                    if (this.hasInvalidStyle(key, attr.value)) {
                        this.report(node, reportMessage(key, attr.value, this.options.style), attr.location);
                    }
                }
            });
        });
    }
    isBoolean(key, rules) {
        return rules[key] && rules[key].length === 0;
    }
}
function parseStyle(style) {
    switch (style.toLowerCase()) {
        case 'omit': return (key, value) => typeof value !== 'undefined';
        case 'empty': return (key, value) => value !== '';
        case 'name': return (key, value) => value !== key;
        default:
            throw new Error(`Invalid style "${style}" for "attribute-boolean-style" rule`);
    }
}
function reportMessage(key, value, style) {
    switch (style.toLowerCase()) {
        case 'omit': return `Attribute "${key}" should omit value`;
        case 'empty': return `Attribute "${key}" value should be empty string`;
        case 'name': return `Attribute "${key}" should be set to ${key}="${key}"`;
    }
    return '';
}
module.exports = AttributeBooleanStyle;
