// html-validate/build/context/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ContentModel;
(function (ContentModel) {
    ContentModel[ContentModel["TEXT"] = 1] = "TEXT";
    ContentModel[ContentModel["SCRIPT"] = 2] = "SCRIPT";
})(ContentModel = exports.ContentModel || (exports.ContentModel = {}));
class Context {
    constructor(source) {
        this.state = undefined;
        this.string = source.data;
        this.filename = source.filename;
        this.offset = 0;
        this.line = source.line;
        this.column = source.column;
        this.contentModel = ContentModel.TEXT;
        this.scriptEnd = undefined;
    }
    consume(n, state) {
        if (typeof n !== 'number') {
            n = n[0].length;
        }
        let consumed = this.string.slice(0, n);
        let offset;
        while ((offset = consumed.indexOf('\n')) >= 0) {
            this.line++;
            this.column = 1;
            consumed = consumed.substr(offset + 1);
        }
        this.column += consumed.length;
        this.offset += n;
        this.string = this.string.substr(n);
        if (typeof state !== 'undefined') {
            this.state = state;
        }
    }
    getLocation(size) {
        return {
            filename: this.filename,
            offset: this.offset,
            line: this.line,
            column: this.column,
            size,
        };
    }
}
exports.Context = Context;
exports.default = Context;
