// html-validate/build/rule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("./config");
class Rule {
    constructor(options) {
        this.options = options;
        this.enabled = true;
    }
    getSeverity() {
        return this.severity;
    }
    setServerity(severity) {
        this.severity = severity;
    }
    setEnabled(enabled) {
        this.enabled = enabled;
    }
    isEnabled() {
        return this.enabled && this.severity >= config_1.Config.SEVERITY_WARN;
    }
    report(node, message, location) {
        if (this.isEnabled()) {
            const where = this.findLocation({ node, location, event: this.event });
            this.reporter.add(node, this, message, this.severity, where);
        }
    }
    findLocation(src) {
        if (src.location) {
            return src.location;
        }
        if (src.event && src.event.location) {
            return src.event.location;
        }
        if (src.node && src.node.location) {
            return src.node.location;
        }
        return {};
    }
    on(event, callback) {
        this.parser.on(event, (event, data) => {
            if (this.isEnabled()) {
                this.event = data;
                callback(data);
            }
        });
    }
    init(parser, reporter, severity) {
        this.parser = parser;
        this.reporter = reporter;
        this.severity = severity;
    }
}
exports.Rule = Rule;
